Packages/com.unity.inputsystem/Tests/InputSystem/Plugins/XRTests.cs

```diff
@@ -1,29 +1,648 @@
+    {
+        testRuntime.ReportNewInputDevice(
+            new InputDeviceDescription
+        {
+            interfaceName = XRUtilities.kXRInterface,
+            product = "XRDevice",
+            manufacturer = "__Manufacturer::",
+            capabilities = new XRDeviceDescriptor
+            {
+                deviceRole = DeviceRole.Generic,
+                inputFeatures = new List<XRFeatureDescriptor>()
+                {
+                    new XRFeatureDescriptor()
+                    {
+                        name = "SimpleFeature[|.:+-?<1",
+                        featureType = FeatureType.Binary
+                    }
+                }
+            }.ToJson()
+        }.ToJson());
+
+        InputSystem.Update();
+
+        var generatedLayout = InputSystem.TryLoadLayout("XRInput::Manufacturer::XRDevice");
+        Assert.That(generatedLayout, Is.Not.EqualTo(null));
+        Assert.That(generatedLayout.controls.Count, Is.EqualTo(1));
+
+        var childControl = generatedLayout.controls[0];
+        Assert.That(childControl.name, Is.EqualTo(new InternedString("SimpleFeature1")));
+    }
+
+    [Test]
+    [Category("Layouts")]
+    public void Layouts_XRDevicesWithNoOrInvalidCapabilitiesDoNotCreateLayouts()
+    {
+        testRuntime.ReportNewInputDevice(
+            new InputDeviceDescription
+        {
+            interfaceName = XRUtilities.kXRInterface,
+            product = "XRDevice1",
+            manufacturer = "XRManufacturer",
+            capabilities = null
+        }.ToJson());
+
+        testRuntime.ReportNewInputDevice(
+            new InputDeviceDescription
+        {
+            interfaceName = XRUtilities.kXRInterface,
+            product = "XRDevice2",
+            manufacturer = "XRManufacturer",
+            capabilities = "Not A JSON String"
+        }.ToJson());
+
+        InputSystem.Update();
+
+        var generatedLayout = InputSystem.TryLoadLayout("XRInput::XRManufacturer::XRDevice1");
+        Assert.That(generatedLayout, Is.EqualTo(null));
+
+        generatedLayout = InputSystem.TryLoadLayout("XRInput::XRManufacturer::XRDevice2");
+        Assert.That(generatedLayout, Is.EqualTo(null));
+    }
+
+    [Test]
+    [Category("Layouts")]
+    public void Layouts_WhenMatchingAKnownDeviceTheGeneratedLayoutInheritsProperly()
+    {
+        testRuntime.ReportNewInputDevice(
+            new InputDeviceDescription
+        {
+            interfaceName = XRUtilities.kXRInterface,
+            product = "Oculus Rift",
+            manufacturer = "Oculus",
+            capabilities = new XRDeviceDescriptor
+            {
+                deviceRole = DeviceRole.LeftHanded,
+                inputFeatures = new List<XRFeatureDescriptor>()
+                {
+                    new XRFeatureDescriptor()
+                    {
+                        name = "SimpleFeature",
+                        featureType = FeatureType.Binary
+                    }
+                }
```

Comment:
This feature doesn't seem relevant to the test. Same for some other tests here.

Reply:
I need a device state of at least 1 element, since it needs to have a size > 0 in the device state.
Although, that reminds me... I should have a test and feature that excludes devices of 0 features.